src/utils/archive.js

```diff
@@ -0,0 +1,116 @@
+const fs = require('fs');
+const path = require('path');
+const archiver = require('archiver');
+const JSZip = require('jszip');
+const isElectron = require('./Environment').isElectron;
+const isCordova = require('./Environment').isCordova;
+const getEnvironment = require('./Environment').getEnvironment;
+
+
+// const zlibFastestCompression = 1;
+// const zlibBestCompression = 9;
+const zlibDefaultCompression = -1;
+
+// Use zlib default: compromise speed & size
+// archiver overrides zlib's default (with 'best speed'), so we need to provide it
+const archiveOptions = {
+  zlib: { level: zlibDefaultCompression },
+  store: true,
+};
+
+/**
+ * Write a bundled (zip) from source files
+ * @param {string} destinationPath full FS path to write
+ * @param {string[]} sourcePaths
+ * @return Returns a promise that resolves to (sourcePath, destinationPath)
+ */
+const archiveElectron = (sourcePaths, destinationPath) =>
+  new Promise((resolve, reject) => {
+    const output = fs.createWriteStream(destinationPath);
+    const zip = archiver('zip', archiveOptions);
+
+    output.on('close', () => {
+      resolve(destinationPath);
+    });
+
+    output.on('warning', reject);
+    output.on('error', reject);
+
+    zip.pipe(output);
+
+    zip.on('warning', reject);
+    zip.on('error', reject);
+
+    sourcePaths.forEach((sourcePath) => {
+      zip.append(fs.createReadStream(sourcePath), { name: path.basename(sourcePath) });
+    });
+
+    zip.finalize();
+  });
+
+const getFile = (filename, fileSystem) => new Promise((resolve, reject) => {
+  fileSystem.root.getFile(filename, { create: false, exclusive: false },
+    fileEntry => resolve(fileEntry),
+    err => reject(err),
+  );
+});
+
+const createReader = fileEntry => new Promise((resolve, reject) => {
+  fileEntry.file(
+    file => resolve(file),
+    err => reject(err),
+  );
+});
+
+/**
+ * Write a bundled (zip) from source files
+ * @param {string} destinationPath full FS path to write
+ * @param {string[]} sourcePaths
+ * @return Returns a promise that resolves to (sourcePath, destinationPath)
+ */
+const archiveCordova = (sourcePaths, targetFileName, fileWriter, filesystem) => {
```

Comment:
To me this looks like it might run more efficiently if we start with the zip promise, and then write the files one at time as we read them. Rather than 1. Read _all_ the files into memory 2. Write all the files from memory?

I'm not sure about how zip works internally, so that may not be the case?

Reply:
I can see if I can reverse this, thought I think it is reading one file and zipping it before doing the next.

Reply:
We need to be mindful of memory presure on cordova too. In general I would take any performance hit that kept memory use low.

Reply:
@rebeccamadsen Hmm you're right, this is writing the files to the zip object one at a time. Looks like since we can't use streams the zip itself has to be generated as a single blob.

Reply:
It looks like there is an option for [`generateAsync`](https://stuk.github.io/jszip/documentation/api_jszip/generate_async.html) to 'streamFiles', but that some programs may not be able to read the zip file. I haven't tried it yet because that sounded worse to me, but maybe we can revisit if cordova memory is a bigger issue than some programs not being able to read the zip? I have it using the highest compression, so hopefully that helps.

Electron does use streaming, using the same method as Server.

Reply:
@rebeccamadsen - what did you decide on this? I looked into the compatibility of generateAsync, and I think the tradeoff is worth it. All native ZIP apps on the major OS's support it. We need the memory reduction on cordova for scenarios where there are many sessions to export.